File: python/176.py
"""Project Euler Problem 176: Right-angled triangles that share a cathetus."""

from typing import List
import math

TARGET = 47_547
VALUE = 2 * TARGET + 1
ODD_PRIMES = [p for p in [3, 5, 7, 11, 13, 17, 19] if p > 2]


def prime_division(value: int) -> List[tuple[int, int]]:
    """Compute prime factorization."""
    factors: List[tuple[int, int]] = []
    d = 2
    while d * d <= value:
        if value % d == 0:
            count = 0
            while value % d == 0:
                value //= d
                count += 1
            factors.append((d, count))
        d += 1
    if value > 1:
        factors.append((value, 1))
    return factors


def divisors_of(value: int) -> List[int]:
    """Generate all divisors of value."""
    factors = prime_division(value)
    divisors = [1]
    for prime, exp in factors:
        current = []
        for e in range(1, exp + 1):
            factor = prime ** e
            for d in divisors:
                current.append(d * factor)
        divisors.extend(current)
    return sorted(set(divisors))


def minimal_m_for(f: int) -> int:
    """Compute the minimal odd component m associated with a given F."""
    factors = prime_division(f)
    exponents: List[int] = []
    for prime, multiplicity in factors:
        for _ in range(multiplicity):
            exponents.append((prime - 1) // 2)
    exponents.sort(reverse=True)
    m = 1
    for idx, exp in enumerate(exponents):
        m *= ODD_PRIMES[idx] ** exp
    return m


def count_triples(n: int) -> int:
    """Count triples for verification."""
    value = n
    a = 0
    while (value & 1) == 0:
        value >>= 1
        a += 1
    odd_part = value
    factors = prime_division(odd_part)
    factor_product = 1
    for _, exp in factors:
        factor_product *= 2 * exp + 1
    if a <= 1:
        return (factor_product - 1) // 2
    else:
        return ((2 * a - 1) * factor_product - 1) // 2


def main() -> int:
    """Main function."""
    best_n: int | None = None

    # Handle a = 0 (odd n) and a = 1 (exactly one factor of 2)
    for a in [0, 1]:
        f = VALUE
        m = minimal_m_for(f)
        n = (1 << a) * m
        if best_n is None or n < best_n:
            best_n = n

    for d in divisors_of(VALUE):
        if d < 3:
            continue
        if (d + 1) % 2 != 0:
            continue
        a = (d + 1) // 2
        if a < 2:
            continue
        f = VALUE // d
        m = minimal_m_for(f)
        n = (1 << a) * m
        if best_n is None or n < best_n:
            best_n = n

    if best_n is None:
        raise ValueError("No solution found")
    return best_n


if __name__ == "__main__":
    result = main()
    print(result)

File: c/176.c
/*
 * Project Euler Problem 176: Right-angled triangles sharing a cathetus.
 *
 * Find the smallest n such that the number of right triangles with
 * n as a cathetus is exactly 47547.
 *
 * The count of right triangles with cathetus n is:
 *   For n = 2^a * m (m odd), the count is:
 *     (prod (2*e_i + 1) - 1) / 2      if a <= 1
 *     ((2a-1) * prod(2*e_i+1) - 1) / 2 if a >= 2
 *   where m = prod p_i^e_i.
 *
 * We need count = 47547, so 2*47547+1 = 95095 = product of odd factors.
 * 95095 = 5 * 7 * 11 * 13 * 19
 *
 * For a <= 1: we need prod(2*e_i + 1) = 95095
 * For a >= 2: we need (2a-1) * prod(2*e_i + 1) = 95095
 *
 * Factor 95095 as product of odd numbers >= 3 (or 1 for the trivial factor).
 * The exponents e_i = (f_i - 1) / 2 where f_i are the factors.
 * To minimize n, assign largest exponents to smallest primes.
 */
#include <stdio.h>
#include <stdlib.h>
#include <string.h>

typedef long long i64;
typedef unsigned long long u64;

#define VALUE 95095  /* 2 * 47547 + 1 */

static int odd_primes[] = {3, 5, 7, 11, 13, 17, 19, 23, 29, 31};
static int n_odd_primes = 10;

/* Prime factorization */
typedef struct { int prime; int exp; } Factor;

static int factorize(int n, Factor *factors) {
    int nf = 0;
    for (int d = 2; (i64)d * d <= n; d++) {
        if (n % d == 0) {
            factors[nf].prime = d;
            factors[nf].exp = 0;
            while (n % d == 0) { n /= d; factors[nf].exp++; }
            nf++;
        }
    }
    if (n > 1) { factors[nf].prime = n; factors[nf].exp = 1; nf++; }
    return nf;
}

/* Get all divisors of n */
static int get_divisors(int n, int *divs) {
    Factor factors[20];
    int nf = factorize(n, factors);
    divs[0] = 1;
    int nd = 1;
    for (int i = 0; i < nf; i++) {
        int prev_nd = nd;
        int pk = 1;
        for (int e = 1; e <= factors[i].exp; e++) {
            pk *= factors[i].prime;
            for (int j = 0; j < prev_nd; j++)
                divs[nd++] = divs[j] * pk;
        }
    }
    return nd;
}

/* Compute minimal m for a given f = product of (2e_i + 1) factors.
 * f's prime factorization gives us the exponent structure.
 * Each prime factor p with multiplicity k contributes k exponents of (p-1)/2.
 * Assign largest exponents to smallest odd primes to minimize m. */
static u64 minimal_m_for(int f) {
    Factor factors[20];
    int nf = factorize(f, factors);

    int exponents[50];
    int ne = 0;
    for (int i = 0; i < nf; i++)
        for (int j = 0; j < factors[i].exp; j++)
            exponents[ne++] = (factors[i].prime - 1) / 2;

    /* Sort descending */
    for (int i = 0; i < ne - 1; i++)
        for (int j = i + 1; j < ne; j++)
            if (exponents[i] < exponents[j]) {
                int tmp = exponents[i]; exponents[i] = exponents[j]; exponents[j] = tmp;
            }

    u64 m = 1;
    for (int i = 0; i < ne; i++) {
        u64 p = odd_primes[i];
        for (int e = 0; e < exponents[i]; e++)
            m *= p;
    }
    return m;
}

int main(void) {
    u64 best = (u64)-1;

    /* Case a = 0 or a = 1: need product = VALUE */
    for (int a = 0; a <= 1; a++) {
        u64 m = minimal_m_for(VALUE);
        u64 n = (1ULL << a) * m;
        if (n < best) best = n;
    }

    /* Case a >= 2: need (2a-1) * prod = VALUE, so (2a-1) | VALUE and (2a-1) >= 3 */
    int divs[500];
    int nd = get_divisors(VALUE, divs);

    for (int di = 0; di < nd; di++) {
        int d = divs[di];
        if (d < 3) continue;
        if ((d + 1) % 2 != 0) continue;
        int a = (d + 1) / 2;
        if (a < 2) continue;
        if (a >= 64) continue;  /* would overflow u64 */
        int f = VALUE / d;
        u64 m = minimal_m_for(f);
        u64 power = 1ULL << a;
        if (m > 0 && power > (u64)-1 / m) continue;  /* mul overflow */
        u64 n = power * m;
        if (n < best) best = n;
    }

    printf("%llu\n", best);
    return 0;
}
